tests/js/client/shell/transaction/replication2_recovery/shell-transaction-replication2-recovery.js

```diff
@@ -0,0 +1,276 @@
+    print(e.message);
+  });
+};
+
+/**
+ * In this test suite we check if the DocumentState can survive modifications to the cluster participants
+ * during transactions. We check for failover and moveshard.
+ */
+function transactionReplication2Recovery() {
+  'use strict';
+  const dbn = 'UnitTestsTransactionDatabase';
+  const cn = 'UnitTestsTransaction';
+  const coordinator = replicatedLogsHelper.coordinators[0];
+  let c = null;
+  let shards = null;
+  let shardId = null;
+  let logId = null;
+
+  const { setUpAll, tearDownAll, stopServerWait, continueServerWait, setUpAnd, tearDownAnd } =
+    replicatedLogsHelper.testHelperFunctions(dbn, { replicationVersion: "2" });
+
+  return {
+    setUpAll,
+    tearDownAll,
+    setUp: setUpAnd(() => {
+      c = db._create(cn, { "numberOfShards": 1, "writeConcern": 2, "replicationFactor": 3 });
+      shards = c.shards();
+      shardId = shards[0];
+      logId = shardId.slice(1);
+    }),
+    tearDown: tearDownAnd(() => {
+      if (c !== null) {
+        c.drop();
+      }
+      c = null;
+    }),
+
+    testFailoverDuringTransaction: function () {
+      // Start a transaction.
+      let trx = db._createTransaction({
+        collections: { write: c.name() }
+      });
+      let tc = trx.collection(c.name());
+      tc.insert({ _key: 'test1', value: 1 });
+      tc.insert({ _key: 'test2', value: 2 });
+
+      // Stop the leader. This triggers a failover.
+      const logs = replicatedLogsHttpHelper.listLogs(coordinator, dbn).result;
+      const participants = logs[logId];
+      assertTrue(participants !== undefined);
+      const leader = participants[0];
+      const followers = participants.slice(1);
+      let term = replicatedLogsHelper.readReplicatedLogAgency(dbn, logId).plan.currentTerm.term;
+      let newTerm = term + 2;
+
+      // We unset the leader here so that once the old leader node is resumed we
+      // do not move leadership back to that node.
+      replicatedLogsHelper.unsetLeader(dbn, logId);
+
+      stopServerWait(leader);
+      replicatedLogsHelper.waitFor(replicatedLogsPredicates.replicatedLogLeaderEstablished(
+        dbn, logId, newTerm, followers));
+
+      // Check if the universal abort command appears in the log during the current term.
+      let logContents = replicatedLogsHelper.dumpShardLog(shardId);
+      let abortAllEntryFound = _.some(logContents, entry => {
+        if (entry.logTerm !== newTerm || entry.payload === undefined) {
+          return false;
+        }
+        return entry.payload[1].operation === "AbortAllOngoingTrx";
+      });
+      assertTrue(abortAllEntryFound);
+
+      // Expect further transaction operations to fail.
+      try {
+        tc.insert({ _key: 'test3', value: 3 });
+        fail('Insert was expected to fail due to transaction abort.');
+      } catch (ex) {
+        assertEqual(internal.errors.ERROR_TRANSACTION_NOT_FOUND.code, ex.errorNum);
+      }
+
+      syncShardsWithLogs(dbn);
+
```

Comment:
I would like to add a check here that. validates:
1) test1 is not to be found on any server
2) test2 is not to be found on any server
2) test3 is not to be found on any server

I would actually like to see this test 3 times:
First: After transaction start, before killing the leader (should be isolated up to that point), if possible i would like to sync the shards before killing the leader, then run this test on all servers, then kill the leader.

Second: After the transaction officially failed

Third: After the leader rejoined. It should not have it's own garbage lying around somewhere.

Reply:
I updated the test accordingly.